Development HomeworkStudent

QuestroL opened the pull request:
1. Create 2 new entity : HomeworkStudent (this entity for Student . He can use it for him homework what his have done) , AttachementOfHomeworkStudent 
2. Create 2 Dto for HomeworkStudentController
3. add 4 Endpoint in HomeworkStudentController (Post, Put Get(for stident), Get (For Mentro))

QuestroL committed:
first changes

QuestroL committed:
secend generation

QuestroL committed:
Merge with dev

QuestroL committed:
add using

QuestroL committed:
merge

andy-dendy-repo commented on CharlieBackend.Data/Repositories/Impl/LessonRepository.cs:
```diff
@@ -33,6 +33,12 @@ public async Task<List<Lesson>> GetAllLessonsForMentor(long mentorId)
                 .ToListAsync();
         }
 
+        public async Task<Lesson> GetLessonByHomeworkId(long homeworkId)
+        {
+            return await _applicationContext.Lessons
+                .FirstOrDefaultAsync(l => l.Id == _applicationContext.Homeworks.FirstOrDefault(h => h.Id == homeworkId).LessonId);
+        }
+
```
This method searches through all Lessons and Homeworks to find a pair. I think, it's better to get Homework by ID and then get LessonId. Having LessonId, you can get Lesson by ID.

int LessonID = _applicationContext.Homeworks.FirstOrDefault(h => h.Id == homeworkId).LessonId;

Lesson lesson = await _applicationContext.Lessons.FirstOrDefaultAsync(l =>l.Id == LessonID);

QuestroL replied:
@andy-dendy-repo you have writed the same code . i made only one Select,  you made two Select,  nothing is change .  Maybe i dont understand something.

mahotskuy replied:
Try to make the next call _applicationContext.Homeworks.FirstOrDefault(h => h.Id == homeworkId).Include(Lesson).Select(homework => homework.Lesson)

QuestroL replied:
@mahotskuy  it doesnt work

Hlib-mats commented on CharlieBackend.Core/DTO/Homework/HomeworkRequestDto.cs:
```diff
@@ -9,6 +9,7 @@ public class HomeworkRequestDto
     {
         public DateTime? DueDate { get; set; }
 
+        [Required]
```
Does TaskText actualy must be required? Homework may be without any explanation

QuestroL replied:
@Hlib-mats yes , we need this field. it cannt be like you writed, because mentor must write task for student and TaskTest exactly for this .

Hlib-mats replied:
okay, persuasively

Hlib-mats commented on CharlieBackend.Core/DTO/HomeworkStudent/HomeworkStudentRequestDto.cs:
```diff
@@ -0,0 +1,18 @@
+        [Required]
+        public long HomeworkId { get; set; }
+       
+        [Required]
```
Also here homework may be without any text

QuestroL replied:
@Hlib-mats  yes, because we have attachment and we can upload home tast into blockstorage . so i think we dont need write something in this field, but we need make validation that both field is not empty.

Hlib-mats commented on CharlieBackend.Business/Services/HomeworkStudentService.cs:
```diff
@@ -0,0 +1,199 @@
+            _logger = logger;
+        }
+
+        public async Task<Result<HomeworkStudentDto>> CreateHomeworkFromStudentAsync(HomeworkStudentRequestDto homeworkStudent, ClaimsPrincipal userContext)
```
I think here u should remove ClaimsPrincipal and add CurrentUserService in future

QuestroL replied:
I already did it

gingerbread-monster commented on CharlieBackend.Business/Services/HomeworkStudentService.cs:
```diff
@@ -0,0 +1,199 @@
+
+        public async Task<Result<HomeworkStudentDto>> UpdateHomeworkFromStudentAsync(HomeworkStudentRequestDto homeworkStudent, ClaimsPrincipal userContext, long homeworkId)
+        {
+            long accountId = Convert.ToInt32(userContext.Claims.First(x => x.Type.EndsWith("AccountId")).Value);
```
Please, update using `ICurrentUserService`

QuestroL replied:
I have done this

QuestroL committed:
some change

QuestroL committed:
Merge branch 'dev' into feature/addHomeworkStudent

QuestroL committed:
final change

QuestroL committed:
change script for Db

QuestroL committed:
fix conflict

QuestroL committed:
fix bug

QuestroL committed:
fix db script

gingerbread-monster commented on CharlieBackend.Api/Controllers/HomeworkStudentController.cs:
```diff
@@ -0,0 +1,92 @@
+        /// 2.HomeworkTest - here Student can write his Homework or something else
+        /// 3.Attacment(LIst) -   here must be Id of Attachment where student uploaded materials for his Homework</param>
+        [SwaggerResponse(200, type: typeof(HomeworkStudentDto))]
+        [Authorize(Roles = "Student")] //?? admin 
```
```suggestion
        [Authorize(Roles = "Student")]
```
Judging by this line's comment you're not sure if admin should have access to these endpoints or not.
But in your implementation of `CreateHomeworkFromStudentAsync()` you are using `_currentUserService.AccountId`, which no one but `Student` would have a valid value for.

So I think, despite the fact that admin can access all other endpoints (see the [endpoints' accessibility](https://github.com/ita-social-projects/WhatBackend/wiki/Endpoint-accessibility) page), we can leave it as is.

gingerbread-monster commented on CharlieBackend.Api/Controllers/HomeworkStudentController.cs:
```diff
@@ -0,0 +1,92 @@
+        /// </summary>
+        /// <param name="request">
+        /// 1.HomeworkId -  it is id homework which one student done 
+        /// 2.HomeworkTest - here Student can write his Homework or something else
```
```suggestion
        /// 2.HomeworkText - here Student can write his Homework or something else
```
what do you mean by *something else* ? :D

QuestroL replied:
@gingerbread-monster  i dont know

gingerbread-monster commented on CharlieBackend.Business/Services/Interfaces/IHomeworkStudentService.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace CharlieBackend.Business.Services.Interfaces
+{
+    public interface IHomeworkStudentService
```
Perhaps we should be adding summaries so there would be less confusion between `HomeworkStudentService` and `HomeworkService`.

gingerbread-monster commented on CharlieBackend.Business/Services/Interfaces/IHomeworkStudentService.cs:
```diff
@@ -0,0 +1,19 @@
+    {
+        Task<Result<HomeworkStudentDto>> CreateHomeworkFromStudentAsync(HomeworkStudentRequestDto homeworkStudent);
+
+        Task<IList<HomeworkStudentDto>> GetHomeworkStudentForMentorByHomeworkId(long homeworkId);
+
+        Task<IList<HomeworkStudentDto>> GetHomeworkStudentForStudent();
```
```suggestion
        Task<IList<HomeworkStudentDto>> GetHomeworkForMentor(long homeworkId);

        Task<IList<HomeworkStudentDto>> GetHomeworkForStudent();
```
Inconsistency: in one case you're adding suffix `ByHomeworkId`, in other case you don't. I think it is redundant.

gingerbread-monster commented on CharlieBackend.Data/Repositories/Impl/Interfaces/IHomeworkStudentRepository.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        Task<bool> IsStudentHasHomeworkAsync(long studentId, long homeworkId);
+
+        Task<IList<HomeworkStudent>> GetHomeworkStudentForStudentByStudentId(long id);
+
+        Task<IList<HomeworkStudent>> GetHomeworkStudentForMentorByHomeworkId(long homeworkId);
```
*student... for student by student*
Please simplify these method names

gingerbread-monster commented on CharlieBackend.Data/Repositories/Impl/HomeworkStudentRepositrory.cs:
```diff
@@ -0,0 +1,67 @@
+                .FirstOrDefaultAsync(x => x.Id == id);
+        }
+
+        //this method check if the student has already done homework for task(homeworkId)
```
Please, either remove this comment or add it in a summary.

gingerbread-monster commented on CharlieBackend.Business/Services/HomeworkStudentService.cs:
```diff
@@ -0,0 +1,215 @@
+            return _mapper.Map<IList<HomeworkStudentDto>>(homeworkStudent);
+        }
+
+        private async IAsyncEnumerable<string> ValidateHomeworStudentRequest(HomeworkStudentRequestDto homeworkStudent, Student student , Homework homework)
```
```suggestion
        private async IAsyncEnumerable<string> ValidateHomeworkStudentRequest(HomeworkStudentRequestDto homeworkStudent, Student student , Homework homework)
```
typo

gingerbread-monster commented on CharlieBackend.Business/Services/HomeworkStudentService.cs:
```diff
@@ -0,0 +1,215 @@
+
+            await _unitOfWork.CommitAsync();
+
+            _logger.LogInformation($"Homework with id {newHomework.Id} for student with id {accountId} has been added ");
```
Why are we logging this information?

QuestroL committed:
Update CharlieBackend.Api/Controllers/HomeworkStudentController.cs

Co-authored-by: gingerbread-monster <38729881+gingerbread-monster@users.noreply.github.com>

QuestroL committed:
Rework GetLessonByHomeworkId in LessonRepository
Add description For HomeworkStudentRepository methods
Change name of some Methods

QuestroL committed:
Merge branch 'feature/addHomeworkStudent' of https://github.com/ita-social-projects/WhatBackend into feature/addHomeworkStudent

QuestroL committed:
fix error

QuestroL committed:
Update CharlieBackend.Api/Controllers/HomeworkStudentController.cs

Co-authored-by: gingerbread-monster <38729881+gingerbread-monster@users.noreply.github.com>

QuestroL committed:
Update CharlieBackend.Business/Services/HomeworkStudentService.cs

Co-authored-by: gingerbread-monster <38729881+gingerbread-monster@users.noreply.github.com>

andy-dendy-repo commented on CharlieBackend.Data/Repositories/Impl/LessonRepository.cs:
```diff
@@ -33,6 +33,14 @@ public async Task<List<Lesson>> GetAllLessonsForMentor(long mentorId)
 
+        public async Task<Lesson> GetLessonByHomeworkId(long homeworkId)
+        {
+            var homework = await _applicationContext.Homeworks.FirstOrDefaultAsync(x => x.Id == homeworkId);
```
Maybe its better to take LessonId at once
```
var lessonId= await _applicationContext.Homeworks.FirstOrDefaultAsync(x => x.Id == homeworkId).LessonId;

return await _applicationContext.Lessons
                .FirstOrDefaultAsync(l => l.Id == lessonId);
```

andy-dendy-repo commented on CharlieBackend.Business/Services/HomeworkStudentService.cs:
```diff
@@ -0,0 +1,210 @@
+            return Result<HomeworkStudentDto>.GetSuccess(_mapper.Map<HomeworkStudentDto>(foundStudentHomework));
+        }
+
+        public async Task<IList<HomeworkStudentDto>> GetHomeworkStudentForStudent()
```
It seems to be too complex name. GetStudentHomeworks may match better.

gingerbread-monster commented on CharlieBackend.Business/Services/HomeworkStudentService.cs:
```diff
@@ -0,0 +1,215 @@
+
+namespace CharlieBackend.Business.Services
+{
+    public class HomeworkStudentService : IHomeworkStudentService
```
I suggest adding summaries to `HomeworkStudentService` and `HomeworkService` classes in order to avoid confusion in the future. By reading summary we should be able to tell the difference between the services and understand their purpose.

Perhaps something like this
```suggestion
    /// <summary>
    /// Service that provides methods for student's actions with Homework
    /// </summary>
    public class HomeworkStudentService : IHomeworkStudentService
```

And then summary for `HomeworkService` could look something like this:
```cs
    /// <summary>
    /// Service that provides methods for administrative and mentor's work 
    /// with Homeworks
    /// </summary>
    public class HomeworkService : IHomeworkService
```

QuestroL committed:
add description for HomeworkService and HomeworkStudentService

Hlib-mats commented:
LGTM!
